src/elf.c

```diff
@@ -63,6 +67,21 @@ void elf_generate_header(void)
     }
 
     elf32_hdr_t hdr;
+    int phnum = 2, shnum = 8, shstrndx = 7;
+    int shoff = elf_header_len + elf_code->size + elf_data->size +
+                elf_rodata->size + elf_symtab->size + elf_strtab->size +
+                elf_shstr->size;
+    if (dynlink) {
+        phnum += 2;
+        shnum += 7;
+        shstrndx += 7;
+        shoff +=
+            dynamic_sections.elf_interp->size +
+            dynamic_sections.elf_relplt->size + dynamic_sections.elf_plt->size +
+            dynamic_sections.elf_got->size + dynamic_sections.elf_dynstr->size +
+            dynamic_sections.elf_dynsym->size +
+            dynamic_sections.elf_dynamic->size;
+    }
```

Comment:
Problems:
- Magic numbers `2` and `7` lack comments or constant definitions
- Calculation logic scattered across multiple locations, prone to inconsistency